[Clang] Implement CWG2813: Class member access with prvalues

MitalAshok opened the pull request:
[CWG2813](https://cplusplus.github.io/CWG/issues/2813.html)

`prvalue.member_fn(expression-list)` now will not materialize a temporary for `prvalue` if `member_fn` is an explicit object member function, and `prvalue` will bind directly to the object parameter.

The `E1` in `E1.static_member` is now a discarded-value expression, so if `E1` was a call to a `[[nodiscard]]` function, there will now be a warning. This also affects C++98 with `[[gnu::warn_unused_result]]` functions.

This should not affect C where `TemporaryMaterializationConversion` is a no-op.

Closes #100314, fixes #100341

MitalAshok committed:
[Clang] Implement CWG2813

MitalAshok committed:
fix clangd AST test

MitalAshok committed:
Run [[nodiscard]] test in C++23

Sirraide commented on clang/test/CXX/dcl.dcl/dcl.attr/dcl.attr.nodiscard/p2.cpp:
```diff
@@ -124,21 +125,67 @@ void usage() {
+  void explicit_object_member_function(this X self);
 #endif
+};
+
```
This is missing tests for MS properties (though I’d maybe put those in a different file).

MitalAshok commented on clang/lib/Sema/SemaExprMember.cpp:
```diff
@@ -1140,26 +1131,68 @@ Sema::BuildMemberReferenceExpr(Expr *BaseExpr, QualType BaseExprType,
 
-  if (MSPropertyDecl *PD = dyn_cast<MSPropertyDecl>(MemberDecl))
+  if (MSPropertyDecl *PD = dyn_cast<MSPropertyDecl>(MemberDecl)) {
+    // Properties treated as non-static data members for the purpose of
+    // temporary materialization
+    if (MakeGLValue())
```
(Tested locally on v19.41.33923) MSVC does not seem to implement CWG2813 so this:

```c++
struct S {
    S(const S&) = delete;
    S() = default;
    int getX(this S) { return 42; }
    __declspec(property(get = getX))
    int x;
};

int main() {
    return S().x;
}
```

... doesn't compile (attempts to call `S::S(const S&)` on the temporary to construct the explicit object argument).

I assume MSVC will eventually allow this, so we should wait to materialize the temporary here? Especially since https://learn.microsoft.com/en-gb/cpp/cpp/property-cpp?view=msvc-170 describes properties directly as a transformation from data member access -> member function call. I'll get this changed

Sirraide replied:
Yeah, I think this is a case where, even though MSVC doesn’t allow this at the moment, I personally think that it would be in accordance with the intent behind this DR to allow this here (though, of course, it is a compiler extension, so the MSVC people can decide to do whatever they want to here). 

From what I can tell, it also isn’t a breaking change: it’s just that some code that previously wouldn’t compile now does, but please correct me if I’m wrong about this.

Also, I’d maybe mention in a comment that we’re intentionally diverging from MSVC’s behaviour here and why.

MitalAshok commented on clang/lib/Sema/SemaExprMember.cpp:
```diff
@@ -1174,6 +1207,9 @@ Sema::BuildMemberReferenceExpr(Expr *BaseExpr, QualType BaseExprType,
       valueKind = VK_PRValue;
       type = Context.BoundMemberTy;
```
For non-static member function, we don't require a glvalue anymore (to allow by-value explicit objects to bind to prvalues)

Sirraide replied:
Oh yeah, duh, makes sense.

MitalAshok commented on clang/lib/Sema/SemaStmt.cpp:
```diff
@@ -387,9 +388,16 @@ void Sema::DiagnoseUnusedExprResult(const Stmt *S, unsigned DiagID) {
+  if (DiagID == diag::warn_unused_comma_left_operand && isSFINAEContext())
+    return;
+
+  // Don't diagnose discarded left of dot in static class member access
+  // because its type is "used" to determine the class to access
+  if (OrigDiagID == diag::warn_discarded_class_member_access)
+    return;
```
This code path is called for all static member accesses, not just on prvalues.

The problem is in stuff like `discarded.static_member_fn();`. This function is called for `(discarded, E2)` and `discarded;` (expression-statement), where those were previously the only two discarded-value expressions. These had an easy fix of casting to `void`, but that doesn't work for this new discarded-value. https://godbolt.org/z/9b949rvxc

I think that this warning shouldn't be applied when the thing that is discarded is/was dependently typed. Maybe for `s.x()` if we have a declaration `S s;` we can suggest `S::x()`, but that shouldn't be on by default. Or `S().x()` into `S::x()` if that constructor call has no side-effects. Maybe that is more appropriate for clang-tidy than the compiler though?

For now this PR doesn't introduce any new diagnostics, just existing warnings about [[nodiscard]] values being discarded in new positions. I think I'll just split some of `Sema::DiagnoseUnusedExprResult` into `Sema::DiagnoseDiscardedValue` for specifically checking [[nodiscard]] stuff and call that instead without a new diagnostic

Sirraide replied:
> I think that this warning shouldn't be applied when the thing that is discarded is/was dependently typed.

Yeah, I think that makes sense.

> existing warnings about `[[nodiscard]]` values being discarded in new positions.

That’s what I would expect, so that also makes sense.

MitalAshok committed:
Change __declspec(property) handling; Refactor code; Tests

Create Sema::DiagnoseDiscardedNodiscard
Remove warn_discarded_class_member_access and call Sema::DiagnoseDiscardedNodiscard instead
Remove MakeGLValue from the MSPropertyRefExpr path
Fix [[nodiscard]] test in test/CXX
Add test for [[nodiscard]] and explicit object member functions in test/SemaCXX

Sirraide commented on clang/lib/Sema/SemaStmt.cpp:
```diff
@@ -222,18 +222,13 @@ static bool DiagnoseNoDiscard(Sema &S, const WarnUnusedResultAttr *A,
-  const Expr *E = dyn_cast_or_null<Expr>(S);
-  if (!E)
-    return;
+namespace {
```
For functions, we typically mark them as `static` rather than using an anonymous namespace.

Sirraide commented on clang/lib/Sema/SemaStmt.cpp:
```diff
@@ -222,18 +222,13 @@ static bool DiagnoseNoDiscard(Sema &S, const WarnUnusedResultAttr *A,
-  if (!E)
-    return;
+namespace {
+void DiagnoseUnused(Sema &S, const Expr *E, std::optional<unsigned> DiagID) {
+  bool NoDiscardOnly = !DiagID.has_value();
```
A comment explaining the relationship between the `DiagID` parameter and `NoDiscardOnly` here would be nice so anyone looking at calls to this knows what’s going on w/o having to go through the entire function.

Sirraide commented on clang/test/CXX/dcl.dcl/dcl.attr/dcl.attr.nodiscard/p2.cpp:
```diff
@@ -171,21 +187,3 @@ void f() {
-// cxx11-warning@87 {{use of the 'nodiscard' attribute is a C++20 extension}}
-// cxx11-warning@91 {{use of the 'nodiscard' attribute is a C++17 extension}}
-// cxx11-warning@92 {{use of the 'nodiscard' attribute is a C++20 extension}}
-// cxx11-warning@95 {{use of the 'nodiscard' attribute is a C++20 extension}}
```
I will say, this was very confusing to me just now. Are there any actual changes left in this file at this point? This refactoring here is nice, but it should be part of a separate NFC patch if there are no other changes in this file...

Sirraide reviewed (approved):
A few more minor things, but LGTM otherwise.

MitalAshok committed:
anonymous namespace -> static

MitalAshok committed:
Merge branch 'main' into prvalue-explicit-object-member-function-call

MitalAshok committed:
Merge branch 'main' into prvalue-explicit-object-member-function-call

MitalAshok committed:
Update cwg28xx.cpp

MitalAshok committed:
Merge branch 'main' into prvalue-explicit-object-member-function-call

MitalAshok committed:
Fix test failure

MitalAshok committed:
Merge branch 'main' into prvalue-explicit-object-member-function-call

MitalAshok committed:
Retarget clang 20

MitalAshok committed:
Add example from #100341

cor3ntin commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -10673,6 +10673,11 @@ class Sema final : public SemaBase {
+  /// DiagnoseDiscardedNodiscard - Given an expression that is semantically
+  /// a discarded-value expression, diagnose if any [[nodiscard]] value
+  /// has been discarded
+  void DiagnoseDiscardedNodiscard(const Expr *E);
```
Can we find a better name for that function?

cor3ntin commented on clang/lib/Sema/SemaStmt.cpp:
```diff
@@ -222,17 +222,17 @@ static bool DiagnoseNoDiscard(Sema &S, const WarnUnusedResultAttr *A,
-    return;
+static void DiagnoseUnused(Sema &S, const Expr *E,
+                           std::optional<unsigned> DiagID) {
+  // When called from Sema::DiagnoseUnusedExprResult, DiagID is a diagnostic for
+  // where this expression is not used. When called from
+  // Sema::DiagnoseDiscardedNodiscard, DiagID is std::nullopt and this function
+  // will only diagnose [[nodiscard]], [[gnu::warn_unused_result]] and similar
```
```suggestion
  // Diagnoses unused expressions that call functions marked [[nodiscard]], [[gnu::warn_unused_result]] and similar. Additionally, a DiagID can be provided to emit a warning in additional contexts (such as for an unused LHS of a comma expression)
```

cor3ntin replied:
TBH, `warn_unused_comma_left_operand` seem to be the only scenario at a glance so maybe we would be better off introduce an enum for that 3rd parameter

cor3ntin replied:
@MitalAshok ping :)

cor3ntin replied:
@MitalAshok ping again :)

cor3ntin reviewed (commented):
General direction looks good

h-vetinari commented on clang/www/cxx_dr_status.html:
```diff
@@ -16694,7 +16694,7 @@ <h2 id="cxxdr">C++ defect report implementation status</h2>
     <td>DR</td>
     <td>Class member access with prvalues</td>
-    <td class="unknown" align="center">Unknown</td>
+    <td class="unreleased" align="center">Clang 20</td>
```
Given the lack of activity here, this looks to be in danger of slipping - branch is in about a month (going by the llvm 18 [schedule](https://discourse.llvm.org/t/llvm-18-release-schedule/76175))

cor3ntin replied:
Yeah, if @MitalAshok can't push that through the finish line, Maybe one of us need to take over
